src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs

```diff
@@ -115,12 +134,77 @@ private static bool IsValidForGenericConstraintContext(ISymbol symbol)
+                return namedType.GetTypeMembers().Any(static m => IsValidForGenericConstraintContext(m));
+
+            return true;
+        }
+
+        private static bool IsValidForInheritanceContext(ISymbol symbol, INamedTypeSymbol inheritingFrom, TSyntaxContext context)
+        {
+            if (symbol is IAliasSymbol alias)
+                symbol = alias.Target;
+
+            // Technically there is a possibility that a namespace contains no valid elements.
+            // But the profit of removing it from the completion list is pretty low while the cost of a possibly large graph walk can be quite significant,
+            // so we just allow any namespace to be present
+            if (symbol.IsNamespace())
+                return true;
+
+            if (symbol is not INamedTypeSymbol namedType ||
+                symbol.IsDelegateType() ||
+                namedType.IsEnumType())
             {
-                return namedType.GetTypeMembers().Any(IsValidForGenericConstraintContext);
+                return false;
             }
 
-            return true;
+            if (namedType.IsStructType() || namedType.IsModuleType())
+                return namedType.GetTypeMembers().Any(m => IsValidForInheritanceContext(m, inheritingFrom, context));
+
+            // Case when checking a class/record in a valid class/record context
+            // Not applicatable when inheriting from struct, because structs cannot be inherited from class/record
+            if ((context.IsBaseClassContext || context.IsBaseRecordContext) &&
+                namedType.TypeKind is TypeKind.Class &&
+                inheritingFrom.TypeKind is not TypeKind.Struct)
+            {
+                // Static and sealed types are not valid for inheritance context,
+                // but they can have valid nested members (e.g. public class nested in a static class),
+                // so we need to check that
+                if (namedType.IsStatic || namedType.IsSealed)
+                    return namedType.GetTypeMembers().Any(m => IsValidForInheritanceContext(m, inheritingFrom, context));
+
+                // These types are special. They are not sealed, but trying to inherit from them
+                // produces compiler error both in C# and VB
+                if (namedType.SpecialType is
+                        SpecialType.System_Enum or
+                        SpecialType.System_ValueType or
+                        SpecialType.System_Delegate or
+                        SpecialType.System_Array)
+                {
+                    return false;
+                }
+
+                // Is current class is equal to one we inherit from
+                // then such self-reference is only valid if there are interfaces nested in that type
+                if (namedType.Equals(inheritingFrom))
+                    return namedType.GetTypeMembers().Any(m => m.IsInterfaceType());
```

Comment:
this has to be recursive right.  you could have `class C : C.NestedClass.NestedInterface`?